### PR TITLE
Look for assets relative to the binary path

### PR DESCRIPTION
**Pull request by link2xt**
Follow-up to #117

### CODE REVIEW

**Comment by adbenitez** [comment]
> Probably building assets into the binary and unpacking them to blobdir is a proper solution.

+1, lets then leave the current approach of using ./assets in the cwd until we pack the assets in the binary in the future

**Comment by adbenitez** [comment]
> I made it work by looking in the current directory if no assets are found near the binary.

nice!

**Review by adbenitez — APPROVED** [review]

File: src/utils.rs [hunk 1]
```diff
@@ -39,6 +38,24 @@ pub async fn configure_from_env(ctx: &Context) -> Result<()> {
...
+        .parent()
+        .with_context(|| format!("cannot determine binary directory for {}", exe.display()))?
+        .join("assets");
+    if dir.exists() {
+        Ok(dir)
+    } else {
+        Ok(PathBuf::from("assets"))
+    }
+}
+
```
Comment (Septias) [inline src/utils.rs:58]: I personally like to use `#[cfg(debug_assertions)]` to make such decisions. Breaks running `cargo r --release` but that should be fine imo.

File: src/utils.rs [hunk 2]
```diff
@@ -164,16 +181,22 @@ pub enum Webxdc {
...
 
-    pub fn get_path(&self) -> PathBuf {
-        PathBuf::from(self.get_str_path())
+    pub fn get_str_path(&self) -> Result<String> {
+        let path = self.get_path()?;
+        let s = path
+            .to_str()
+            .with_context(|| format!("cannot convert path {} to string", path.display()))?;
+        Ok(s.to_string())
     }
```
Comment (Septias) [inline src/utils.rs:200]:
```suggestion
        self.get_path()?
            .to_str()
            .with_context(|| format!("cannot convert path {} to string", path.display()))
            .map(|str| str.to_string())
    }
```
For me this way nice, but that of cause is personal preference

Comment (Septias) [reply src/utils.rs:200]: Might even be possible to further simplify the map with something like `to_owned()`

File: src/utils.rs [hunk 3]
```diff
@@ -209,20 +232,17 @@ impl WebxdcVersions {
...
-        .iter()
-        .all(|path| path.try_exists().unwrap_or_default());
-
-    if !required_files_present {
-        bail!("It seems like the frontend hasn't been build yet! Look at the readme for further instructions.")
+    for webxdc in Webxdc::iter() {
+        let webxdc_path = webxdc.get_path().context("cannot get webxdc path")?;
+        if !webxdc_path.try_exists()? {
+            bail!("Required webxdc {} is not found.", webxdc_path.display());
+        }
```
Comment (Septias) [inline src/utils.rs:239]: very nice!

**Review by Septias — COMMENTED** [review]
The whole python test refactoring looks amazing, very nice :)

**Review by hpk42 — APPROVED** [review]
lgtm, let's merge.

**Comment by Septias** [comment]
@hpk42 would like to have @link2xt look over this before merging:
https://github.com/webxdc/store/pull/119#discussion_r1246481790

**Pull request merged**